### PR TITLE
Feature/intercept robot

### PR DESCRIPTION
**Pull request by rolfvdhulst**
Tries to intercept the path of a robot.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -192,7 +192,7 @@ set(SKILLS_SOURCES
         ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/DemoAttack.cpp
         ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/DemoAttack.h
         ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/SideAttacker.cpp
-        ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/SideAttacker.h)
+        ${PROJECT_SOURCE_DIR}/roboteam_ai/src/skills/SideAttacker.h roboteam_ai/src/skills/InterceptRobot.cpp roboteam_ai/src/skills/InterceptRobot.hpp)
```
Comment (mrlukasbos) [inline CMakeLists.txt]: project source dir

File: roboteam_ai/src/skills/InterceptRobot.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    if (!robotToIntercept){
+        return Status::Failure;
+    }
+    Vector2 robotVel=robotToIntercept->vel;
+    double maxVel=2.0;
+    double maxDist=1.0;
+    double minDist=2.5*Constants::ROBOT_RADIUS();
+    Vector2 interceptPos=robotToIntercept->pos;
+    // we stand at a point in front of the robot depending on it's speed
+    // psst don't tell mechanics
```
Comment (mrlukasbos) [inline roboteam_ai/src/skills/InterceptRobot.cpp]: useless comment

File: roboteam_ai/src/skills/InterceptRobot.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    Vector2 angle=robotToIntercept->angle.toVector2(1.0);
+    if (robotVel.length()<maxVel){
+        interceptPos+=angle.stretchToLength(minDist+robotVel.length()*(maxDist-minDist));
+    }
+    else{
+        interceptPos+=angle.stretchToLength(maxDist);
+    }
+    control::PosVelAngle velocities=gtp.getPosVelAngle(robot,interceptPos);
```
Comment (mrlukasbos) [inline roboteam_ai/src/skills/InterceptRobot.cpp]: can probably split this in a function or two

File: roboteam_ai/src/skills/InterceptRobot.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        :Skill(name, blackboard) {
+
+}
+Skill::Status InterceptRobot::onUpdate() {
+    RobotPtr robotToIntercept=world::world->getRobotForId(1,true);
+    if (!robotToIntercept){
+        return Status::Failure;
+    }
+    Vector2 robotVel=robotToIntercept->vel;
+    double maxVel=2.0;
```
Comment (mrlukasbos) [inline roboteam_ai/src/skills/InterceptRobot.cpp]: magic numbers

File: roboteam_ai/src/skills/InterceptRobot.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    control::PosVelAngle velocities=gtp.getPosVelAngle(robot,interceptPos);
+    Vector2 XYvel=control::ControlUtils::velocityLimiter(velocities.vel,4.0);
+    command.x_vel=XYvel.x;
+    command.y_vel=XYvel.y;
+    command.w=velocities.angle.getAngle();
+
+    publishRobotCommand();
+    return Status::Running;
+}
+void InterceptRobot::onInitialize() {
```
Comment (mrlukasbos) [inline roboteam_ai/src/skills/InterceptRobot.cpp]: can you put the initialize above the onupdate? ;-)

File: roboteam_ai/src/skills/InterceptRobot.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    gtp.setAvoidBall(false);
+    gtp.setCanMoveInDefenseArea(false);
+    gtp.setCanMoveOutOfField(false);
+}
+void InterceptRobot::onTerminate(Skill::Status s) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/skills/InterceptRobot.cpp]: if you don't use this functino you can remove it

File: roboteam_ai/src/skills/InterceptRobot.hpp [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+namespace rtt{
+namespace ai{
+class InterceptRobot : public Skill{
+    private:
+        control::BasicPosControl gtp;
+    public:
+        explicit InterceptRobot(string name, bt::Blackboard::Ptr blackboard);
+        Status onUpdate() override;
+        void onInitialize() override;
+        void onTerminate(Status s) override;
```
Comment (mrlukasbos) [inline roboteam_ai/src/skills/InterceptRobot.hpp]: onterminate is not used so doesnt need an override

File: CMakeLists.txt (hunk 1 above)
Comment (rolfvdhulst) [reply CMakeLists.txt]: done.

File: roboteam_ai/src/skills/InterceptRobot.cpp (hunk 5 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/skills/InterceptRobot.cpp]: done

File: roboteam_ai/src/skills/InterceptRobot.cpp (hunk 3 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/skills/InterceptRobot.cpp]: done

**Review by mrlukasbos — APPROVED** [review]
Looking nice!

**Pull request merged**